futures-intrusive/src/sync/mutex.rs

```diff
@@ -0,0 +1,542 @@
+            .finish()
+    }
+}
+
+impl<'a, T> Future for LocalMutexLockFuture<'a, T> {
+    type Output = LocalMutexGuard<'a, T>;
+
+    fn poll(
+        self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> Poll<Self::Output> {
+        // Safety: The next operations are safe, because Pin promises us that
+        // the address of the wait queue entry inside MutexLocalFuture is stable,
+        // and we don't move any fields inside the future until it gets dropped.
+        let mut_self: &mut LocalMutexLockFuture<T> = unsafe {
+            Pin::get_unchecked_mut(self)
+        };
+
+        let mutex = mut_self.mutex.expect("polled LocalMutexLockFuture after completion");
+        let mut mutex_state = mutex.state.borrow_mut();
+
+        let poll_res = unsafe {
+            mutex_state.try_lock(
+            &mut mut_self.wait_node,
+            lw)
+        };
+
+        match poll_res {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(()) => {
+                // The mutex was acquired
+                mut_self.mutex = None;
+                Poll::Ready(LocalMutexGuard::<'a, T>{
+                    mutex,
+                })
+            },
+        }
+    }
+}
+
+impl<'a, T> FusedFuture for LocalMutexLockFuture<'a, T> {
+   fn is_terminated(&self) -> bool {
+       self.mutex.is_none()
+   }
+}
+
+impl<'a, T> Drop for LocalMutexLockFuture<'a, T> {
+    fn drop(&mut self) {
+        // If this LocalMutexLockFuture has been polled and it was added to the
+        // wait queue at the mutex, it must be removed before dropping.
+        // Otherwise the mutex would access invalid memory.
+        if let Some(mutex) = self.mutex {
+            let mut mutex_state = mutex.state.borrow_mut();
+            mutex_state.remove_waiter(&mut self.wait_node);
+        }
+    }
+}
+
+/// A futures-aware mutex.
+pub struct LocalMutex<T> {
+    value: UnsafeCell<T>,
+    state: RefCell<MutexState>,
+}
+
+// It is safe to send mutexes between threads, as long as they are not used and
+// thereby borrowed
+unsafe impl<T: Send> Send for LocalMutex<T> {}
+
+impl<T: core::fmt::Debug> core::fmt::Debug for LocalMutex<T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("LocalMutex")
+            .field("is_locked", &format_args!("{}", self.is_locked()))
+            .finish()
+    }
+}
+
+impl<T> LocalMutex<T> {
+    /// Creates a new futures-aware mutex.
+    ///
+    /// `is_fair` defines whether the `Mutex` should behave be fair regarding the
+    /// order of waiters. A fair `Mutex` will only allow the first waiter which
+    /// tried to lock but failed to lock the `Mutex` once it's available again.
+    /// Other waiters must wait until either this locking attempt completes, and
+    /// the `Mutex` gets unlocked again, or until the `MutexLockFuture` which
+    /// tried to gain the lock is dropped.
+    pub fn new(value: T, is_fair: bool) -> LocalMutex<T> {
```

Comment:
Could be also 2 Mutex types (e.g. `FairLocalMutex`) instead of a parameterized one. Or one version could go away.
I felt that adding a parameter might make it easier to change the behavior in an application later on when required, und reduces duplication in the code. But open to all suggestions